Adding training scripts for two chest_x-ray_screening

Rakshith2597 opened the pull request:
Submitting training module for two models for chest x-ray screening. The first model employs DenseNet-121 CNN architecture. The second CNN model was obtained by performing a systematic search over a set of CNN architectures obtained by scaling the width and depth of DenseNet using the EfficientNet method.

This is part of the project MIRIAD: Many Incarnations of Screening of Radiology for High Throughput Disease Screening via Multiple Instance Reinforcement Learning with Adversarial Deep Neural Networks, sponsored by INTEL TECHNOLOGY INDIA PVT. LTD.

**Principal Investigators:**

Dr Debdoot Sheet (PI), Dr Nirmalya Ghosh (Co-PI)
Department of Electrical Engineering
Indian Institute of Technology Kharagpur

Dr Ramanathan Sethuraman (Co-PI)
Intel Technology India Pvt. Ltd.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+    #---- dataPRED - predicted data
+    #---- classCount - number of classes
+    
+    def computeAUROC (dataGT, dataPRED, classCount):
```
You define this function but you don't use it. On line 78 you use the same function defined in `RSNATrainer` class but you do not import this class. So it would be better to extract this function into separate file and import it from there to `Trainer` and `Test` classes.

Rakshith2597 replied:
Created a separate function outside the class.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+    def test(model, dataLoaderTest, nnClassCount, checkpoint, class_names,device):   
+        cudnn.benchmark = True
+        
+        if checkpoint != None:
+            modelCheckpoint = torch.load(checkpoint)
+            model.load_state_dict(modelCheckpoint['state_dict'])
+        else:
+            model.state_dict()
```
Same as above (in `train.py` file). Don't compare with singletons (like `None`) via comparison operators. Extract loading model into separate function and import it everywhere it is used.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+import time
+import sys
+import csv
+import argparse
+import torch
+from torchvision import models
+import torch.nn as nn
+import torch.backends.cudnn as cudnn
+import torchvision
+import torchvision.transforms as transforms
+import torch.optim as optim
+import torch.nn.functional as tfunc
+from torch.utils.data import Dataset
+from torch.utils.data.dataset import random_split
+from torch.utils.data import DataLoader
+from torch.optim.lr_scheduler import ReduceLROnPlateau,StepLR
+from PIL import Image
+import torch.nn.functional as func
+import random
+from sklearn.metrics.ranking import roc_auc_score
+import sklearn.metrics as metrics
+import json
+from pathlib import Path
+from dataloader import RSNADataSet
```
Please review if all imports are needed and delete unused.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+        aurocIndividual = RSNATrainer.computeAUROC(tfunc.one_hot(outGT.squeeze(1).long()).float(), outPRED, nnClassCount)
+        aurocMean = np.array(aurocIndividual).mean()
+        
+        print ('\nAUROC mean ', aurocMean)
```
Please use the same way of formatting strings everywhere.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+    class_names = ['Lung Opacity','Normal','No Lung Opacity / Not Normal']
+
+    ############# Data Loader ####################
+    img_pth='/home/rakshith/RSNA_pneumonia/dataset/rsna-pneumonia-detection-challenge/train_data/'
```
Local path

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+    model = model.to(device)
+
+    ############## Train the  Model #################
+    timestampTime = time.strftime("%H%M%S")
+    timestampDate = time.strftime("%d%m%Y")
+    timestampLaunch = timestampDate + '-' + timestampTime
```
Same as above about timestamp.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/test.py:
```diff
@@ -0,0 +1,135 @@
+    
+    args = parser.parse_args()
+
+    main(args)
```
No newline in the end of the file.

morkovka1337 reviewed (changes requested):
Overall comments applicable to every part of the code: check if all imports are really needed, correctly use OOP in Python, extract mutual parts of the code in functions instead of writing it in several parts of the code, don't use magic numbers (define names constants and use them instead), don't use local paths.

By the way, what is the purpose of `chest_xray_screening_optimised`? Could this code be merged with `chest_xray_screening`? When you merge it, please, take into account what I've wrote above (magic numbers, code organization, etc).

Use automatic tools for formatting code (`autopep8`) and code analyzers (like `pylint`).

Please, check if all requirements from adding your model [guide](https://github.com/openvinotoolkit/training_extensions/blob/develop/CONTRIBUTING.md#adding-training-code-of-the-model) are met. Particularly, provide bash script `init_venv.sh`. It is used for users to automatically create environment and by testing system to test code.
Don't forget to write tests for the model. Tests should cover the main scenarios of use your model: train, eval, export. In export tests check if both converted (ONNX and OpenvinoIR) models has the same accuracy as source (pytorch model).

You can take a look at other models training code to look at the tests functionality.

Rakshith2597 commented on misc/pytorch_toolkit/chest_xray_screening/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
```
Okay, I Will remove it.

Rakshith2597 commented on misc/pytorch_toolkit/chest_xray_screening/tools/dataloader.py:
```diff
@@ -0,0 +1,56 @@
+import os
```
Removed one of the files.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening_optimised/utils/generate.py:
```diff
@@ -0,0 +1,201 @@
+    return total
+
+
+def densenet_flops(input_shape, block_config, growth_rate,
+                   k, init_features, verbose=1):
+    out1 = (input_shape[0] - 7 + 2 * 3) // 2 + 1
+    out2 = (input_shape[1] - 1) // 2 + 1
+    x = conv2d_flops(out1, out2, init_features, 3, 7)
```
If I understand correctly, this code counts flops for a model. You can use the [ready python module](https://github.com/sovrasov/flops-counter.pytorch/blob/master/ptflops/flops_counter.py) for it.

morkovka1337 reviewed (changes requested):
The previous comment about tests is still applicable:
> **Don't forget to write tests for the model. Tests should cover the main scenarios of use your model: train, eval, export. In export, tests check if both converted (ONNX and OpenvinoIR) models have the same accuracy as the source (PyTorch model).
> You can take a look at other models' training code to look at the functionality of the test.**

Provide some small (toy) dataset for testing the main scripts' functionality. Usually, it is about 5-10 instances.

Rakshith2597 commented:
- Merged two models to a single directory
- Made minor changes requested 
- Toy dataset provided
- Updated unit tests
- Re-organised codes as suggested for the other model (PR#546) and fixed few bugs.
- Fixed pylint issues

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/export.py:
```diff
@@ -0,0 +1,44 @@
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--onnx",
+    required=False,
+    help="Set to True, if you wish to export onnx model",
+    default=False,
+    action='store_true')
+    parser.add_argument("--ir",
+    required=False,
+    help="Set to True, if you wish to export IR",
+    default=False,
+    action='store_true')
+    parser.add_argument("--optimised",
+    required=False,
+    help="Set to True, if you wish to export optimised model",
+    default=False,
+    action='store_true')
```
Please, add indentation (like in `tools/inference.py`)

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/inference.py:
```diff
@@ -0,0 +1,123 @@
+
+if __name__=="__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--checkpoint",
+    required=False,
+    help="start training from a checkpoint model weight",
+    default= None,
+    type = str)
+    parser.add_argument("--dpath",
+    required=True,
+    help="Path to dataset",
+    type =str)
+    parser.add_argument("--optimised",
+    required=False,
+    default=False,
+    help="enable flag for eff model",
+    action='store_true')
```
please, use same indentation, as for the parameters below

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+        self.gepoch_id = 0
+        self.model_val = ''
+        self.device = device
+        # if self.device is 'cuda':
```
```suggestion
```
or, if you really want to compare arguments, use `==` operator, not `is`

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+
+class RSNATrainer():
+
+    def __init__(
+        self, model,
+        data_loader_train, data_loader_valid, data_loader_test,
+        class_count, checkpoint, device, class_names,lr):
```
```suggestion
    def __init__(self, model,
                      data_loader_train, data_loader_valid, data_loader_test,
                      class_count, checkpoint, device, class_names, lr):
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+            self.model_checkpoint = None
+        self.loss_fn = torch.nn.BCELoss()
+        self.lr = lr
+        self.optimizer = optim.Adam(self.model.parameters(),lr=self.lr)
```
```suggestion
        self.optimizer = optim.Adam(self.model.parameters(), lr=self.lr)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+            if valid_loss < valid_loss_min:
+                valid_loss_min = valid_loss
+
+            torch.save({'epoch': epoch_id + 1,
+            'state_dict': self.model.state_dict(),
+            'best_loss': valid_loss_min,
+            'optimizer' : self.optimizer.state_dict()},
+            savepath+'/m-epoch'+str(epoch_id)+'-' + timestamp_launch + '.pth.tar')
```
```suggestion
            torch.save({'epoch': epoch_id + 1,
                             'state_dict': self.model.state_dict(),
                             'best_loss': valid_loss_min,
                             'optimizer' : self.optimizer.state_dict()},
                             savepath+'/m-epoch'+str(epoch_id)+'-' + timestamp_launch + '.pth.tar')
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+        # model = model.to(device)
+    else:
+        model = DenseNet121(class_count)
+        # model = model.to(device)
```
```suggestion
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+        beta = beta ** phi
+
+        model = DenseNet121Eff(alpha,beta,class_count)
+        # model = model.to(device)
```
```suggestion
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+            # Evaluate the performance of the model on validation_set
+            # every 2500th iteration. 2500 is a random choice, this could be changed.
+
+            if batch_id%2500==0 and batch_id!=0:
```
[Magic number](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad). Declare a constant in the head of the file (between the imports and the `class` statement), set it 2500 and refer to it.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/data_prep.py:
```diff
@@ -0,0 +1,106 @@
+    file_list = os.listdir(args.dpath+'/processed_data/')
+    df_class=pd.read_csv(args.dpath+'/stage_2_detailed_class_info.csv')
+    labels=["Lung Opacity","Normal","No Lung Opacity / Not Normal"]
+    # image_name = []
+    # image_label = []
```
```suggestion
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/data_prep.py:
```diff
@@ -0,0 +1,106 @@
+        patient_id = file.split('.jpg')[0]
+        tmp=df_class[df_class["patientId"]==patient_id]["class"].values[0]
+        idx=labels.index(tmp)
+        # image_name.append(file)
+        # image_label.append(idx)
```
```suggestion
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/data_prep.py:
```diff
@@ -0,0 +1,106 @@
+        # image_label.append(idx)
+        dict_annotation[str(file)] = idx
+
+    # dict_annotation = {'names':image_name,'labels':image_label}
```
```suggestion
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tests/test_export.py:
```diff
@@ -0,0 +1,43 @@
+
+    def test_export_ir(self):
+
+        self.assertTrue(os.path.isdir('/opt/intel/openvino_2021'))
```
declare a variable for `/opt/intel/openvino_2021` like in `tools/exporter.py` and refer to it

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/train.py:
```diff
@@ -0,0 +1,280 @@
+                param.requires_grad = True
+            print("Model loaded")
+
+        train_loss_min = 100000  # A random very high number
+        valid_loss_min = 100000
```
```suggestion
        train_loss_min = 1e+5  # A random very high number
        valid_loss_min = 1e+5
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/generate.py:
```diff
@@ -0,0 +1,69 @@
+    width_temp = gw * l4 + (gw * l3 + (gw * l2 + (gw * l1) // 2) // 2) // 2
+    init_chan = int(8 * (1024 * beta - width_temp))
+    while init_chan < 32:
+        print(f"The initial number of channels{init_chan} needs to be greater than 32. Changing growth_rate.")
```
```suggestion
        print(f"The initial number of channels {init_chan} needs to be greater than 32. Changing growth_rate.")

```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/generate.py:
```diff
@@ -0,0 +1,69 @@
+    gw = growth_rate
+    bn_size = round(alpha * bn_size)
+    if bn_size < 2:
+        print(f"The initial bn_size{bn_size} needs to be greater than 2.")
```
```suggestion
        print(f"The initial bn_size {bn_size} needs to be greater than 2.")

```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/generate.py:
```diff
@@ -0,0 +1,69 @@
+
+def get_size(blocks, growth_rate, initial_channels):
+
+    depth = 5 + 2 * sum(blocks)
```
It would be nice to add some description of what 5 and 2 stand for

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/generate.py:
```diff
@@ -0,0 +1,69 @@
+        init_chan = 2
+    width_temp = gw * l4 + (gw * l3 + (gw * l2 + (gw * l1) // 2) // 2) // 2
+    init_chan = int(8 * (1024 * beta - width_temp))
+    while init_chan < 32:
```
magic number

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tools/utils/generate.py:
```diff
@@ -0,0 +1,69 @@
+        growth_rate -= 1
+        gw = growth_rate
+        width_temp = gw * l4 + (gw * l3 + (gw * l2 + (gw * l1) // 2) // 2) // 2
+        init_chan = int(8 * (1024 * beta - width_temp))
```
magic number

morkovka1337 reviewed (changes requested):
Please, 
- [ ] Move toy data to the `data` folder in the root of this repo (create a subdir `chest_x-ray_screening`). 
- [ ] Format the code using some code editor (indentation, spaces after comma, etc). If you use VS Code, hit `Ctrl+Alt+F` when the file is opened, other editors should also support such feature.

morkovka1337 commented:
run tests

morkovka1337 commented:
Add execute permissions for the init_venv.sh script to be able to run it. `sudo chmod +x init_venv.sh`

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/README.md:
```diff
@@ -0,0 +1,242 @@
+</tr>
+</table>
+
+A systematic search was performed over a set of CNN architectures by scaling the width and depth of the standard DenseNet using the EfficientNet approach[[3](efficientnet)]. The details of our experiments based on EfficientNet is summarized in the image below.
```
```suggestion
A systematic search was performed over a set of CNN architectures by scaling the width and depth of the standard DenseNet using the EfficientNet approach[[3](#efficientnet)]. The details of our experiments based on EfficientNet is summarized in the image below.
```

Rakshith2597 commented:
> run tests

Did you mean the unit tests that I have provided in the tests folder?

morkovka1337 commented:
> > run tests
> 
> Did you mean the unit tests that I have provided in the tests folder?

No, this is a trigger command for the jenkins to run the tests ;)

morkovka1337 commented:
run tests

Rakshith2597 commented:
> * [x]  Move toy data to the `data` folder in the root of this repo (create a subdir `chest_x-ray_screening`).
> * [x]  Format the code using some code editor (indentation, spaces after a comma, etc). If you use VS Code, hit `Ctrl+Alt+F` when the file is opened, other editors should also support such a feature.

@morkovka1337  Modified the codes as requested.

morkovka1337 commented:
run tests

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/init_venv.sh:
```diff
@@ -0,0 +1,24 @@
+  echo "$ . venv/bin/activate"
+fi
+
+virtualenv venv -p python3
```
```suggestion
virtualenv ${venv_dir} -p python --prompt="(chest_x-ray_screening)"
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/init_venv.sh:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+work_dir=$(realpath "$(dirname $0)")
+
+cd ${work_dir}
```
```suggestion
work_dir=$(realpath "$(dirname $0)")

venv_dir=$1
if [ -z "$venv_dir" ]; then
  venv_dir=venv
fi

cd ${work_dir}
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/init_venv.sh:
```diff
@@ -0,0 +1,24 @@
+virtualenv venv -p python3
+
+
+. venv/bin/activate
```
```suggestion
path_openvino_vars="${INTEL_OPENVINO_DIR:-/opt/intel/openvino}/bin/setupvars.sh"
if [[ -e "${path_openvino_vars}" ]]; then
  echo ". ${path_openvino_vars}" >>${venv_dir}/bin/activate
fi

. ${venv_dir}/bin/activate
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/init_venv.sh:
```diff
@@ -0,0 +1,24 @@
+cat requirements.txt | xargs -n 1 -L 1 pip3 install
+
+
+echo
+echo "Activate a virtual environment to start working:"
+echo "$ . venv/bin/activate"
```
```suggestion
echo
echo "Activate a virtual environment to start working:"
echo "$ . ${venv_dir}/bin/activate"
```

morkovka1337 reviewed (commented):
CI test passes the `venv_dir` argument, and without it venv is initialized, but in another folder. The test asserts to have venv in the temporary folder, but there is no venv, and the test fails.

morkovka1337 commented:
run tests

Rakshith2597 commented:
> CI test passes the `venv_dir` argument, and without it venv is initialized, but in another folder. The test asserts to have venv in the temporary folder, but there is no venv, and the test fails.

I forgot to update the init_venv file for this model. I hope the changes made by you will fix the issue.

morkovka1337 commented:
Please, rename `tools` to `chest_screening` since `tools` folder is the openvino subfolder, and the test tries to import tools.utils from openvino/tools, that causes the error.

Rakshith2597 commented:
> Please, rename `tools` to `chest_screening` since `tools` folder is the openvino subfolder, and the test tries to import tools.utils from openvino/tools, that causes the error. You will also need to change imports (`from tools.utils import` -> `from chest_screening.utils import`).

Renamed the folders

morkovka1337 commented:
run tests

morkovka1337 commented:
Looks like the problem with imports still exists:
```
23:16:32  ==================================== ERRORS ====================================
23:16:32  ___________________ ERROR collecting tests/test_inference.py ___________________
23:16:32  ImportError while importing test module '/home/jenkins/b/workspace/ote/precommit/ote_tests_runner/src/misc/pytorch_toolkit/chest_xray_screening/tests/test_inference.py'.
23:16:32  Hint: make sure your test modules/packages have valid Python names.
23:16:32  Traceback:
23:16:32  /tmp/tmpkuq0nuth/chest_xray_screening/lib/python3.7/importlib/__init__.py:127: in import_module
23:16:32      return _bootstrap._gcd_import(name[level:], package, level)
23:16:32  tests/test_inference.py:7: in <module>
23:16:32      from chest_xray_screening.utils.data import DataLoader
23:16:32  E   ModuleNotFoundError: No module named 'chest_xray_screening.utils.data'
23:16:32  _____________________ ERROR collecting tests/test_train.py _____________________
23:16:32  ImportError while importing test module '/home/jenkins/b/workspace/ote/precommit/ote_tests_runner/src/misc/pytorch_toolkit/chest_xray_screening/tests/test_train.py'.
23:16:32  Hint: make sure your test modules/packages have valid Python names.
23:16:32  Traceback:
23:16:32  /tmp/tmpkuq0nuth/chest_xray_screening/lib/python3.7/importlib/__init__.py:127: in import_module
23:16:32      return _bootstrap._gcd_import(name[level:], package, level)
23:16:32  tests/test_train.py:4: in <module>
23:16:32      from chest_xray_screening.train import RSNATrainer
23:16:32  chest_xray_screening/train.py:11: in <module>
23:16:32      from utils.dataloader import RSNADataSet
23:16:32  E   ModuleNotFoundError: No module named 'utils'
23:16:32  =========================== short test summary info ============================
23:16:32  ERROR tests/test_inference.py
23:16:32  ERROR tests/test_train.py
23:16:32  !!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!
23:16:32  ============================== 2 errors in 0.72s ===============================
```

morkovka1337 commented:
run tests

Rakshith2597 commented:
> Looks like the problem with imports still exists:

I have changed `utils.dataloader`  to `chest_xray_screening.utils.dataloader`. **utils** folder and **train.py** are located inside the same chest_xray_screening folder. In such a case, the import should have worked with utils.dataloader itself, right? Because in my local system, it works fine.

morkovka1337 commented:
> > Looks like the problem with imports still exists:
> 
> I have changed `utils.dataloader` to `chest_xray_screening.utils.dataloader`. **utils** folder and **train.py** are located inside the same chest_xray_screening folder. In such a case, the import should have worked with utils.dataloader itself, right? Because in my local system, it works fine.

I've tested in my local system too, and it does not work. But if I use relative imports (like `from .utils.dataloader import RSNADataSet`), there is no problem with the imports, but there are other errors related to the fact that there is no checkpoint in the `model_weights` folder. You can try to use relative imports, and download model checkpoint from the GDrive if it does not exists in the path specified in the config. You can create the folder `model_weights` via `os.makedirs` and download the checkpoint using something like [this](https://github.com/ndrplz/google-drive-downloader). When you finish, you can try to pull your repo in some temporary folder and make sure everything works if the model does not exist in the current folder or simply trigger the CI by `run tests` command (as I did above).

morkovka1337 commented:
By the way, how to open the tar with the model? I failed to open it using 7Zip in Windows and tar in Ubuntu. Also, the model name in the config file ends with `.pt`, not `.pth`, possibly, this is a typo. :)

Rakshith2597 commented:
> in the config file ends with `.pt`, not `.pth`, possibly, this is a typo. :)
Yes, that is a typo in this case. We generally save the file in .pt format, so we made a mistake.

> how to open the tar 
In Linux, you can open tar files using the `tar -xvf filename.tar` command on the terminal. There seems to be some error while we extract the model_weights.tar file. I will reupload this file. It is working for other files.

Rakshith2597 commented:
Converting this to a draft. There seems to be a few errors in the tests.

Rakshith2597 commented:
>  Move toy data to the data folder in the root of this repo (create a subdir chest_x-ray_screening).

`data/*` is mentioned in the .gitignore file. Due to this, I'm unable to push the dummy dataset to the data folder at the root of this repo. How should I proceed?

morkovka1337 commented:
Hi, you can add ignored file with the help of the `-f` option: `git add -f filename`.

Rakshith2597 commented:
run tests

Rakshith2597 commented:
run tests

Rakshith2597 commented:
run tests

Rakshith2597 commented:
run tests

Rakshith2597 commented:
run tests

Rakshith2597 commented:
@morkovka1337 Hey, I do not see any errors in the console output. The only thing I was able to identify was 

> Failed: ote_tests_runner http://213.221.44.203/job/ote/job/precommit/job/ote_tests_runner/137/
> Failed: ote_linting http://213.221.44.203/job/ote/job/precommit/job/ote_linting/74/

What does this indicate, and how can I resolve this, If this is the reason for failure?

morkovka1337 commented:
> 
> 
> @morkovka1337 Hey, I do not see any errors in the console output. The only thing I was able to identify was
> 
> > Failed: ote_tests_runner http://213.221.44.203/job/ote/job/precommit/job/ote_tests_runner/137/
> > Failed: ote_linting http://213.221.44.203/job/ote/job/precommit/job/ote_linting/74/
> 
> What does this indicate, and how can I resolve this, If this is the reason for failure?

You can click on the links above and then go to console output. You should see this in the end:
```
02:21:36  ==================================== ERRORS ====================================
02:21:36  ____________________ ERROR collecting tests/test_export.py _____________________
02:21:36  tests/test_export.py:5: in <module>
02:21:36      from ..utils.exporter import Exporter
02:21:36  E   ValueError: attempted relative import beyond top-level package
02:21:36  ___________________ ERROR collecting tests/test_inference.py ___________________
02:21:36  tests/test_inference.py:7: in <module>
02:21:36      from ..utils.dataloader import RSNADataSet
02:21:36  E   ValueError: attempted relative import beyond top-level package
02:21:36  _____________________ ERROR collecting tests/test_train.py _____________________
02:21:36  tests/test_train.py:7: in <module>
02:21:36      from ..chest_xray_screening.train import RSNATrainer
02:21:36  E   ValueError: attempted relative import beyond top-level package
02:21:36  =========================== short test summary info ============================
02:21:36  ERROR tests/test_export.py - ValueError: attempted relative import beyond top...
02:21:36  ERROR tests/test_inference.py - ValueError: attempted relative import beyond ...
02:21:36  ERROR tests/test_train.py - ValueError: attempted relative import beyond top-...
02:21:36  !!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!
02:21:36  ============================== 3 errors in 0.52s ===============================
```

morkovka1337 commented:
The errors in the `ote_linting` are not suppressed and its output could be more useful:
http://213.221.44.203/job/ote/job/precommit/job/ote_linting/74/console
```
02:23:22  ************* Module tests.test_train
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_train.py:7:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_train.py:8:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_train.py:9:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  ************* Module tests.test_export
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_export.py:5:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  ************* Module tests.test_inference
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_inference.py:7:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_inference.py:8:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/tests/test_inference.py:9:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  ************* Module chest_xray_screening.export
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/export.py:3:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  ************* Module chest_xray_screening.inference
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/inference.py:10:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/inference.py:11:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/inference.py:12:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  ************* Module chest_xray_screening.train
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:11:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:12:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
02:23:22  misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:13:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
```

Rakshith2597 commented:
> You can click on the links above and then go to console output. You should see this in the end:

Found it. Thanks.

morkovka1337 commented:
If I'm not mistaken, you can fix it with adding `setup.py` and adding it's call to the `init_venv.sh`:
`pip install -e .`
Then relative imports should work correct.

Rakshith2597 commented:
![Capture](https://user-images.githubusercontent.com/25449712/128979807-5c6b17db-13b2-48b5-a4cc-9285c9a2caae.PNG)
This is the output I am getting on my local server at the moment.

Rakshith2597 commented:
> If I'm not mistaken, you can fix it with adding `setup.py` and adding it's call to the `init_venv.sh`:
> `pip install -e .`
> Then relative imports should work correct.

I have added this now.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:
```diff
@@ -0,0 +1,260 @@
+
+    def epoch_train(self,savepath):
+        auroc_max = 0.0 # Setting maximum AUROC value as zero
+        if self.model_checkpoint is not None:
+            self.optimizer.load_state_dict(self.model_checkpoint['optimizer'])
```
BTW, what is the purpose of loading checkpoint every epoch iteration?

Rakshith2597 replied:
Moved it outside the loop.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:
```diff
@@ -0,0 +1,260 @@
+            loss_train_list.append(train_loss_value)
+
+            if batch_id%len(self.data_loader_train)==0 and batch_id!=0:
+                validloss_value,auroc_mean = RSNATrainer.valid(self)
```
```suggestion
                validloss_value,auroc_mean = self.valid()
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:
```diff
@@ -0,0 +1,260 @@
+        num_workers=4,
+        pin_memory=False)
+
+    dataset_test = RSNADataSet(test_list,labels,img_pth,transform=True)
```
```suggestion
    dataset_test = RSNADataSet(test_list, labels, img_pth, transform=True)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:
```diff
@@ -0,0 +1,260 @@
+
+if __name__=="__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--lr",required=False, help="Learning rate",default=1e-4,type = float)
+    parser.add_argument("--checkpoint",required=False, help="Checkpoint model weight",default= None ,type = str)
+    parser.add_argument("--bs",required=False, default=16, help="Batchsize", type=int)
+    parser.add_argument("--dpath",required=True, help="Path to folder containing all data", type =str)
+    parser.add_argument("--epochs",required=False,default=15, help="Number of epochs", type=int)
+    parser.add_argument("--clscount",required=False,default=3, help="Number of classes", type=int)
+    parser.add_argument("--spath",required=True, help="Path to folder in which models should be saved", type =str)
+    parser.add_argument("--optimised",required=False,default=False,help="enable flag->eff model",action='store_true')
+    parser.add_argument("--alpha",required=False,help="alpha for the model",default=(11 / 6),type=float)
+    parser.add_argument("--phi",required=False,help="Phi for the model.",default=1.0,type=float)
+    parser.add_argument("--beta",required=False,help="Beta for the model.",default=None,type=float)
```
```suggestion
    parser.add_argument("--lr", required=False, help="Learning rate", default=1e-4,type = float)
    parser.add_argument("--checkpoint", required=False, help="Checkpoint model weight", default= None, type = str)
    parser.add_argument("--bs", required=False, default=16, help="Batchsize", type=int)
    parser.add_argument("--dpath", required=True, help="Path to folder containing all data", type =str)
    parser.add_argument("--epochs", required=False, default=15, help="Number of epochs", type=int)
    parser.add_argument("--clscount", required=False, default=3, help="Number of classes", type=int)
    parser.add_argument("--spath", required=True, help="Path to folder in which models should be saved", type =str)
    parser.add_argument("--optimised", required=False, default=False, help="enable flag->eff model", action='store_true')
    parser.add_argument("--alpha", required=False, help="alpha for the model", default=(11 / 6), type=float)
    parser.add_argument("--phi", required=False, help="Phi for the model.", default=1.0, type=float)
    parser.add_argument("--beta", required=False, help="Beta for the model.", default=None, type=float)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/train.py:
```diff
@@ -0,0 +1,260 @@
+        alpha = alpha ** phi
+        beta = beta ** phi
+
+        model = DenseNet121Eff(alpha,beta,class_count)
```
```suggestion
        model = DenseNet121Eff(alpha, beta, class_count)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+
+def dcm_to_jpg(args):
+
+    dirpath = args.dpath + '/original/'
+    savepath = args.dpath +'/processed_data/'
```
```suggestion
    dirpath = os.path.join(args.dpath, 'original/')
    savepath = os.path.join(args.dpath, 'processed_data/')
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+        os.makedirs(savepath)
+    file_list = os.listdir(dirpath)
+    for fpath in tq(file_list):
+        ds = dcmread(dirpath+fpath)
```
```suggestion
        ds = dcmread(os.path.join(dirpath, fpath)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+        np_array = ds.pixel_array
+        im = Image.fromarray(np_array)
+        name = fpath.split('.dcm')[0]
+        im.save(savepath+name+".jpg")
```
```suggestion
        im.save(os.path.join(savepath, name+".jpg"))
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def mean_sd_calculator(args):
+    dirpath = args.dpath + '/original/'
```
```suggestion
    dirpath = os.path.join(args.dpath, 'original/')
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+
+def create_annotation(args):
+
+    file_list = os.listdir(args.dpath+'/processed_data/')
+    df_class=pd.read_csv(args.dpath+'/stage_2_detailed_class_info.csv')
```
```suggestion
    file_list = os.listdir(os.path.join(args.dpath, 'processed_data/'))
    df_class=pd.read_csv(os.path.join(args.dpath, 'stage_2_detailed_class_info.csv')
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+    dict_annotation = {}
+    for file in tq(file_list):
+        patient_id = file.split('.jpg')[0]
+        tmp=df_class[df_class["patientId"]==patient_id]["class"].values[0]
```
```suggestion
        tmp = df_class[df_class["patientId"] == patient_id]["class"].values[0]
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+    for file in tq(file_list):
+        patient_id = file.split('.jpg')[0]
+        tmp=df_class[df_class["patientId"]==patient_id]["class"].values[0]
+        idx=labels.index(tmp)
```
```suggestion
        idx = labels.index(tmp)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+        dict_annotation[str(file)] = idx
+
+
+    with open(args.dpath+'/rsna_annotation.json','w') as f:
```
```suggestion
    with open(os.path.join(args.dpath, 'rsna_annotation.json'), 'w') as f:
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/data_prep.py:
```diff
@@ -0,0 +1,101 @@
+
+
+    with open(args.dpath+'/rsna_annotation.json','w') as f:
+        json.dump(dict_annotation,f)
```
```suggestion
        json.dump(dict_annotation, f)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/exporter.py:
```diff
@@ -0,0 +1,44 @@
+OPENVINO_DIR = '/opt/intel/openvino_2021'
+
+class Exporter:
+    def __init__(self, config,optimised):
```
```suggestion
    def __init__(self, config, optimised):
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/exporter.py:
```diff
@@ -0,0 +1,44 @@
+            self.model = DenseNet121(class_count=3)
+
+        self.model.eval()
+        self.checkpoint = config.get('checkpoint')
+        model_checkpoint = torch.load(self.checkpoint)
+        if model_checkpoint is not None:
+            self.model.load_state_dict(model_checkpoint['state_dict'])
```
Same as `train.py` & `inference.py`

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/generate.py:
```diff
@@ -0,0 +1,87 @@
+    growth_rate = round(beta / alpha * growth_rate)
+    gw = growth_rate
+    bn_size = round(alpha * bn_size)
+    if bn_size < 2: # The bn_size should be atleast 1
```
```suggestion
    if bn_size <= 1: # The bn_size should be atleast 1

```
Though it is the same in this case, such if condition corresponds to its comment more ;)

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/model.py:
```diff
@@ -0,0 +1,24 @@
+            param.requires_grad = False
+        self.model.classifier=nn.Sequential(nn.Linear(1024, class_count), nn.Sigmoid())
+    def forward(self, x):
+        x= self.model(x)
```
```suggestion
        x = self.model(x)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/model.py:
```diff
@@ -0,0 +1,24 @@
+        self.model = give_model(alpha, beta, class_count)
+        self.model = nn.Sequential(self.model, nn.Sigmoid())
+    def forward(self, x):
+        x= self.model(x)
```
```suggestion
        x = self.model(x)
```

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tests/test_train.py:
```diff
@@ -0,0 +1,101 @@
+        config = config_file['train']
+    return config
+
+def download_checkpoint():
+    os.makedirs('model_weights')
+    gdd.download_file_from_google_drive(file_id='1z4HuSVXyD59BHhw93j-BVbx6In1HZQn2',
+                                    dest_path='model_weights/chest_xray_screening.pth',
+                                    unzip=False)
+    gdd.download_file_from_google_drive(file_id='1HUmG-wKRoKYxBdwu0_LX1ascBRmA-z5e',
+                                    dest_path='model_weights/chest_xray_screening_eff.pth',
+                                    unzip=False)
```
same about code duplication

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tests/test_train.py:
```diff
@@ -0,0 +1,101 @@
+                                    unzip=False)
+
+class TrainerTest(unittest.TestCase):
+    config = get_config()
+    class_count = config["clscount"]
+    image_path = '../../data/chest_xray_screening/'
+    learn_rate = config["lr"]
+    tr_list = config["dummy_train_list"]
+    val_list = config["dummy_valid_list"]
+    test_list = config["dummy_test_list"]
+    labels = config["dummy_labels"]
```
same about `setUp()` method

morkovka1337 reviewed (changes requested):
I kindly ask you to use some automatic code formatter since it is very long to align spaces by hand. Also, please, replace all path concatenation as string with os.path.join. It is needed for the correct work on non Unix-like systems (like Windows) and add unzipping model from `.tar` to `.pth` for correct test passing.

Rakshith2597 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/inference.py:
```diff
@@ -0,0 +1,124 @@
+            beta = round(sqrt(2 / alpha), 3)
+        alpha = alpha ** phi
+        beta = beta ** phi
+        model = DenseNet121Eff(alpha,beta,class_count)
```
@morkovka1337  Does the space after `,` have any significance other than formatting/readability of the code?

Ilya-Krylov commented:
run tests

morkovka1337 commented:
run tests

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tests/test_export.py:
```diff
@@ -0,0 +1,46 @@
+from chest_xray_screening.utils.exporter import Exporter, OPENVINO_DIR
+
+
+def _get_config_():
+    path = os.path.dirname(os.path.realpath(__file__))
+    with open(os.path.join(path, 'test_config.json'), ) as f1:
+        config_file = json.load(f1)
+
+    return config_file['export']
```
Please, unify this function with similar functions from `test_train` and `test_inference`, move it into some common file and export it from there.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/download_weights.py:
```diff
@@ -0,0 +1,11 @@
+
+def download_checkpoint():
+    os.makedirs('model_weights')
+    gdd.download_file_from_google_drive(file_id='13KtMdL-RJoTXGRZ8WIdcHkf0XuX9bMMN',
+                                    dest_path='model_weights/chest_xray_screening.pth',
```
I think, it would be better to leave this function as a wrapper for `gdd.download_file_from_google_drive`. Add parameters for destination and file id, store these parameters in the config file of the model.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/utils/exporter.py:
```diff
@@ -0,0 +1,41 @@
+        load_checkpoint(self.model, self.checkpoint)
+
+    def export_model_ir(self):
+        input_model = os.path.join(os.path.split(self.checkpoint)[0], self.config.get('model_name'))
```
```suggestion
        input_model = os.path.join(os.path.split(self.checkpoint)[0], self.config.get('model_name_onnx'))
```
Otherwise the code tries to export "model_weights/chest_xray_screening` which does not exist.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/tests/test_export.py:
```diff
@@ -0,0 +1,46 @@
+import json
+from chest_xray_screening.utils.download_weights import download_checkpoint
+from chest_xray_screening.utils.exporter import Exporter, OPENVINO_DIR
+
+
```
By default, the tests are sorted in the alphabetic order. This means, that `test_export_ir` will be invoked before `test_export_onnx` and it will fail, since onnx model does not exits. 
You can disable this ordering, but unit tests should not depend on the output of the previous tests. Check, if onnx model exists in the `test_export_ir` and if it does not, export it.

morkovka1337 reviewed (changes requested):
Currently, test config is stored in the `tests` folder and export config is stored in the root of the project. Please, 1. create some directory, e.g. `configs` and place these configs into this folder; 2. divide the configs based on the model they describe (e.g., `densnet121` and `densenet121eff`) and create sections in the each config (export, test, train). Also store `file_id` for the GDrive and name of the model in these configs.
As far as I understand, you have two models, but you test only the base model against the export, train and inference. Please, add tests for the other model as well. You can look at how it is done in the text recognition.

Rakshith2597 commented:
run tests

Rakshith2597 commented:
run tests

Rakshith2597 commented:
@morkovka1337 

> [2021-08-20T13:41:29.202Z] FAILED tests/test_export.py::TestInference::test_export_ir - subprocess.Calle...
> [2021-08-20T13:41:29.202Z] FAILED tests/test_export.py::TestInferenceEff::test_export_ir - subprocess.Ca...

Two of the unit tests mentioned above are failing. The error which I was able to identify from the logs is given below,

> >  ----------------------------- Captured stdout call -----------------------------
> [2021-08-20T13:41:29.201Z] DenseNet model created with block_config:[5, 11, 22, 15], growth_rate:34, o/p channels:172, bn_size:4
> [2021-08-20T13:41:29.202Z] Saving model to chest_xray_screening_eff.onnx
> [2021-08-20T13:41:29.202Z] [setupvars.sh] OpenVINO environment initialized
> [2021-08-20T13:41:29.202Z] ----------------------------- Captured stderr call -----------------------------
> [2021-08-20T13:41:29.202Z] Traceback (most recent call last):
> [2021-08-20T13:41:29.202Z]   File "/opt/intel/openvino_2021.4.582/deployment_tools/model_optimizer/mo/main.py", line 22, in <module>
> [2021-08-20T13:41:29.202Z]     from extensions.back.SpecialNodesFinalization import RemoveConstOps, CreateConstNodesReplacement, NormalizeTI
> [2021-08-20T13:41:29.202Z]   File "/opt/intel/openvino/deployment_tools/model_optimizer/extensions/back/SpecialNodesFinalization.py", line 10, in <module>
> [2021-08-20T13:41:29.202Z]     from extensions.back.pass_separator import BackFinish
> [2021-08-20T13:41:29.202Z]   File "/opt/intel/openvino/deployment_tools/model_optimizer/extensions/back/pass_separator.py", line 4, in <module>
> [2021-08-20T13:41:29.202Z]     from mo.back.replacement import BackReplacementPattern
> [2021-08-20T13:41:29.202Z]   File "/opt/intel/openvino/deployment_tools/model_optimizer/mo/back/replacement.py", line 4, in <module>
> [2021-08-20T13:41:29.202Z]     from mo.utils import class_registration
> [2021-08-20T13:41:29.202Z]   File "/opt/intel/openvino/deployment_tools/model_optimizer/mo/utils/class_registration.py", line 8, in <module>
**> [2021-08-20T13:41:29.202Z]     import networkx as nx
> [2021-08-20T13:41:29.202Z] ModuleNotFoundError: No module named 'networkx'**

It would be great if you could take a look at this and guide me on how I should proceed.

morkovka1337 commented:
Hi, I'm on vacation currently, I will look at this in detail in a week from now.

morkovka1337 commented:
The `networkx` module is required for the model optimizer. It should be installed when the requirements for the ONNX part of the model optimizer is installed. It can be done in the `init_venv.sh` script, but the easier way now is to install the openvino as a python module (from the [pypi](https://pypi.org/project/openvino/)): `pip install openvino-dev[onnx]`. Thus, if you add `openvino-dev[onnx]` to the requirements, all the necessary dependencies will be installed automatically. 
Doing this, you will also have to change the export command from 
```
f"""{OPENVINO_DIR}/bin/setupvars.sh && \
        python {OPENVINO_DIR}/deployment_tools/model_optimizer/mo.py \
        --framework onnx \
        --input_model {input_model} \
        --input_shape "{input_shape}" \
        --output_dir {output_dir} \
        --scale_values 'imgs[255]'"""
```

to just
```
f"""mo \
        --framework onnx \
        --input_model {input_model} \
        --input_shape "{input_shape}" \
        --output_dir {output_dir} \
        --scale_values 'imgs[255]'"""
```

Rakshith2597 commented:
`------------------------------ Captured stderr call ------------------------------
[ ERROR ]  -------------------------------------------------
[ ERROR ]  ----------------- INTERNAL ERROR ----------------
[ ERROR ]  Unexpected exception happened.
[ ERROR ]  Please contact Model Optimizer developers and forward the following information:
[ ERROR ]  Exception occurred during running replacer "REPLACEMENT_ID (<class 'extensions.middle.AddMeanScaleValues.AddMeanScaleValues'>)": Attempt to access node None that not in graph
[ ERROR ]  Traceback (most recent call last):
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 276, in apply_transform
    replacer.find_and_replace_pattern(graph)
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/extensions/middle/AddMeanScaleValues.py", line 101, in find_and_replace_pattern
    if Node(graph, node_id) not in input_nodes:
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/graph/graph.py", line 25, in __init__
    assert node in graph, "Attempt to access node {} that not in graph".format(node)
AssertionError: Attempt to access node None that not in graph

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/main.py", line 394, in main
    ret_code = driver(argv)
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/main.py", line 356, in driver
    ret_res = emit_ir(prepare_ir(argv), argv)
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/main.py", line 252, in prepare_ir
    graph = unified_pipeline(argv)
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/pipeline/unified.py", line 17, in unified_pipeline
    class_registration.ClassType.BACK_REPLACER
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 328, in apply_replacements
    apply_replacements_list(graph, replacers_order)
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 318, in apply_replacements_list
    num_transforms=len(replacers_order))
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/logger.py", line 111, in wrapper
    function(*args, **kwargs)
  File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 306, in apply_transform
    )) from err
Exception: Exception occurred during running replacer "REPLACEMENT_ID (<class 'extensions.middle.AddMeanScaleValues.AddMeanScaleValues'>)": Attempt to access node None that not in graph

[ ERROR ]  ---------------- END OF BUG REPORT --------------
[ ERROR ]  -------------------------------------------------`

I am getting the above error while running the tests. I had previously converted the ONNX models to IR using the [DL workbench](https://docs.openvinotoolkit.org/latest/openvino_docs_get_started_get_started_dl_workbench.html).

Rakshith2597 commented:
run tests

morkovka1337 commented:
> 
> 
> I am getting the following error while running the tests. I had previously converted the ONNX models to IR using the [DL workbench](https://docs.openvinotoolkit.org/latest/openvino_docs_get_started_get_started_dl_workbench.html).
> 
> ```
> `------------------------------ Captured stderr call ------------------------------
> [ ERROR ]  -------------------------------------------------
> [ ERROR ]  ----------------- INTERNAL ERROR ----------------
> [ ERROR ]  Unexpected exception happened.
> [ ERROR ]  Please contact Model Optimizer developers and forward the following information:
> [ ERROR ]  Exception occurred during running replacer "REPLACEMENT_ID (<class 'extensions.middle.AddMeanScaleValues.AddMeanScaleValues'>)": Attempt to access node None that not in graph
> [ ERROR ]  Traceback (most recent call last):
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 276, in apply_transform
>     replacer.find_and_replace_pattern(graph)
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/extensions/middle/AddMeanScaleValues.py", line 101, in find_and_replace_pattern
>     if Node(graph, node_id) not in input_nodes:
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/graph/graph.py", line 25, in __init__
>     assert node in graph, "Attempt to access node {} that not in graph".format(node)
> AssertionError: Attempt to access node None that not in graph
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/main.py", line 394, in main
>     ret_code = driver(argv)
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/main.py", line 356, in driver
>     ret_res = emit_ir(prepare_ir(argv), argv)
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/main.py", line 252, in prepare_ir
>     graph = unified_pipeline(argv)
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/pipeline/unified.py", line 17, in unified_pipeline
>     class_registration.ClassType.BACK_REPLACER
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 328, in apply_replacements
>     apply_replacements_list(graph, replacers_order)
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 318, in apply_replacements_list
>     num_transforms=len(replacers_order))
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/logger.py", line 111, in wrapper
>     function(*args, **kwargs)
>   File "/home/rakshith/RSNA_pneumonia/training_extensions/misc/pytorch_toolkit/chest_xray_screening/venv/lib/python3.7/site-packages/mo/mo/utils/class_registration.py", line 306, in apply_transform
>     )) from err
> Exception: Exception occurred during running replacer "REPLACEMENT_ID (<class 'extensions.middle.AddMeanScaleValues.AddMeanScaleValues'>)": Attempt to access node None that not in graph
> 
> [ ERROR ]  ---------------- END OF BUG REPORT --------------
> [ ERROR ]  -------------------------------------------------`
> ```

Looks like the input names should be added, otherwise MO is confused about what is the input.
In the export command:
```
f"""mo \
        --framework onnx \
        --input_model {input_model} \
        --input_shape "{input_shape}" \
        --output_dir {output_dir} \
        --input 'imgs' \
        --scale_values 'imgs[255]'"""
```
In the export to the ONNX:
```
input_names=['imgs']
```

Rakshith2597 commented:
run tests

Rakshith2597 commented:
@morkovka1337 The last build was sucessful. Should I make any more changes?

morkovka1337 reviewed (changes requested):
Please, add tests that exported model (both ONNX and OpenVINO IR) can infer and produces the same accuracy as the source model. You can look at the example in the text recognition.

Rakshith2597 commented:
run tests

Rakshith2597 commented:
> Please, add tests that exported model (both ONNX and OpenVINO IR) can infer and produces the same accuracy as the source model. You can look at the example in the text recognition.

I have added this in the test_inference.py file.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/inference.py:
```diff
@@ -0,0 +1,184 @@
+                to_tensor = transforms.ToTensor()
+                if run_type == 'pytorch':
+                    out = self.model(var_input)
+                elif run_type == 'onnx':
+                    ort_session = onnxruntime.InferenceSession(onnx_checkpoint)
+                    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(var_input)}
```
There is no need to load the model in every iteration on the datasets, it is slow. You can load the model in advance and then run the inference in the loop. Also, it seems like now there are two functions for doing the same: both `test` and `validate_models` can run the pytorch model. 

I guess, the better option in this case is to store the `run_type` as the class variable and create separate method, which encapsulates the logic of the specific model inference.
I.e., you can create two methods (e.g., load_model and inference_model) and, depending on the run_type, they will act differently but produce the same result (loading model and doing one inference step).

Rakshith2597 replied:
> both test and validate_models can run the PyTorch model.
Removed `test` and have made necessary changes.

> you can create two methods (e.g., load_model and inference_model)
I have created separate methods.

morkovka1337 commented:
run tests

morkovka1337 commented:
Ok, I think, now the code looks well. 

- [ ] Please, update sections Code and Directory Organisation and Code Structure in the readme according to the current code structure.

- [ ] Also, please, make all run commands consistent (I mean, in the Run Training section python3 train.py  -> python train.py and so on) and I think, we can finally merge it. :)

Rakshith2597 commented:
>  [x]  Please, update sections Code and Directory Organisation and Code Structure in the readme according to the current code structure.
>  [x]  Also, please, make all run commands consistent (I mean, in the Run Training section python3 train.py  -> python train.py and so on) 

Done.

morkovka1337 commented on misc/pytorch_toolkit/chest_xray_screening/README.md:
```diff
@@ -141,26 +141,25 @@ chest_xray_screening/
-3. `generate.py` optimised model is defined in this file. 
-4. `export.py` generating the onnx representation and Openvino IR of the trained model.
-5. `tests` directory contains minimal unittests and a configuration file.
+1. `train.py` in chest_xray_screening diectory contains the code for training the model.
+2. `inference.py` in chest_xray_screening diectory contains the code for evaluating the model with test set.
+3. `export.py` in chest_xray_screening diectory generating the ONNX and Openvino IR of the trained model.
```
```suggestion
1. `train.py` in chest_xray_screening directory contains the code for training the model.
2. `inference.py` in chest_xray_screening directory contains the code for evaluating the model with test set.
3. `export.py` in chest_xray_screening directory generating the ONNX and Openvino IR of the trained model.
```

morkovka1337 reviewed (commented):
About the sample data (in the `data` folder): if I understand correctly, the samples are from https://www.kaggle.com/c/rsna-pneumonia-detection-challenge/? Do you have any information about the dataset license?

Rakshith2597 commented:
> About the sample data (in the `data` folder): if I understand correctly, the samples are from https://www.kaggle.com/c/rsna-pneumonia-detection-challenge/? Do you have any information about the dataset license?

Licence of this dataset: [link](https://www.rsna.org/-/media/Files/RSNA/Education/AI-resources-and-training/AI-image-challenge/pneumonia-detection-challenge-terms-of-use-and-attribution.ashx?la=en&hash=FF7A635F6DFFAD31A30C8715DFA3B8FC21131543)
The challenge was organised by RSNA and was hosted in Kaggle. More details about the dataset and its license can be found here ([link](https://www.rsna.org/education/ai-resources-and-training/ai-image-challenge/RSNA-Pneumonia-Detection-Challenge-2018))

Rakshith2597 commented:
> And what about the images in the `chest_xray_screening/media`, what is their origin?

The images class0.jpg. class1.jpg and class2.jpg were taken from the RSNA dataset (same as the toy dataset). The image efficientnet.jpg is a custom image created by us for explaining the concept.

morkovka1337 commented:
According to the license clause:
> If you share or re-distribute the data in any form, to meet the attribution requirements described below

We have to meet these requirements:
![image](https://user-images.githubusercontent.com/32310433/131995279-3acf7a35-a9ec-4382-88cb-d928dc6bbb4b.png)

Thus, please, add a link to the dataset (to the official RSNA webpage in addition to the kaggle link) and the citation of the above mentioned paper.

Rakshith2597 commented:
> Thus, please, add a link to the dataset (to the official RSNA webpage in addition to the kaggle link) and the citation of the above mentioned paper.

The dataset name and link were already mentioned in the second paragraph. Additionally, I have provided a link to the dataset license for easier reference in future as well citation of the mentioned paper in the reference section.

Rakshith2597 commented:
@morkovka1337 Is there any more changes that should be made by me?

morkovka1337 commented:
Hi.
Sorry for not replying earlier. We need some more approvals to legally use the part of the dataset in our data folder. I've started the procedure today, expect up to two weeks for the next reply. I will post here if things change.
Sorry for that inconvenience. =(

morkovka1337 commented:
Hi, @Rakshith2597!

We have investigated this question with Sethuraman, Ramanathan and came to the conclusion that it would be better to **download the dataset on the fly and use the part of it to perform tests**.

That is, I kindly ask you to do these things:
1. Delete the data from `data/` folder
2. In tests: 
2.1. Check if there is any data in `/mnt/external_test_data/chest_xray_screening`, if yes, perform tests on the images from there
2.2. If no: create some folder with reasonable name like `data` (relatively to the `chest_xray_screening` folder), download and unzip the data there and run the test on the images over there. 
Note, there is no need to download the data for the second and further tests, if it was downloaded in the first, so, there should be two checks: if the data is in `/mnt/external_test_data/chest_xray_screening` and if the data is in `data`.

Rakshith2597 commented:
run tests

Rakshith2597 commented:
@morkovka1337 I have updated the codes as requested.

morkovka1337 commented:
run tests

morkovka1337 reviewed (changes requested):
1. Please, avoid constant strings defined in the file multiple times. If the string changes in one place, it is likely that a programmer will forget to change it in another. (`'/mnt/external_test_data/chest_xray_screening'`). The better option would be to define them in the config file. Something like `default_image_path` and `image_path` and then get it from the config.
2. Please, check that these paths exist with `os.path.exists`, not `os.path.isdir`. 
3. The same is applicable to the `model_weights`. Check that specific model path exists (not only a folder) with `os.path.exists` and only if it does not exist call `download_weights`. The folder could exist, but if it is empty, the model would not be downloaded with the current implementation.

morkovka1337 commented:
You can ignore the failed CI this time, chest_xray tests have successfully passed.
